### PR TITLE
backend errors refactoring

### PR DESCRIPTION
**Pull request by grishasobol**
Proposal status

### CODE REVIEW

File: core-backend/sandbox/src/runtime.rs [hunk 1]
```diff
@@ -91,47 +93,50 @@ where
...
-                })
-                .map(|_| ReturnValue::Unit)
-        } else {
-            Err(HostError)
-        };
-
-        self.update_globals();
-
-        res
+        let write_res = self.memory_manager.register_write_as::<R>(res_ptr);
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core-backend/sandbox/src/runtime.rs]: Done

File: core-backend/common/src/lib.rs [hunk 2]
```diff
@@ -308,81 +224,49 @@ where
 }
 
 #[derive(Debug, Clone, derive_more::From)]
-pub enum SyscallFuncError<E: Display> {
-    Actor(ActorSyscallFuncError<E>),
-    System(SystemSyscallFuncError),
+pub enum FuncError<E: BackendExtError> {
+    #[from]
+    Core(E),
+    #[from]
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core-backend/common/src/lib.rs]: Done

File: core-backend/sandbox/src/runtime.rs [hunk 3]
```diff
@@ -91,47 +93,50 @@ where
...
+                    self.termination_reason = reason;
+                    Err(HostError)
+                }
+            },
+            Ok(res) => Ok(Ok(res)),
+        }?;
+
+        self.write_as(write_res, R::from(res))
+            .map_err(|err| {
+                self.termination_reason = Into::<FuncError<E::Error>>::into(err).into();
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core-backend/sandbox/src/runtime.rs]: Done

File: core-backend/common/src/memory.rs [hunk 4]
```diff
@@ -41,47 +41,19 @@ use gear_core_errors::MemoryError;
...
-
-#[derive(Debug, Clone, derive_more::Display, derive_more::From)]
-pub enum ActorMemoryAccessError {
     #[from]
-    #[display(fmt = "{_0}")]
     Memory(MemoryError),
     #[from]
-    #[display(fmt = "{_0}")]
     RuntimeBuffer(RuntimeBufferSizeError),
+    Decode,
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core-backend/common/src/memory.rs:51]: Done

File: core-errors/src/lib.rs [hunk 5]
```diff
@@ -66,7 +66,7 @@ pub enum MessageError {
     /// send more than the maximum amount of messages allowed within a single
     /// execution (current setting - 1024).
     #[display(fmt = "Message limit exceeded")]
-    LimitExceeded,
+    MessagesNumberLimitExceeded,
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core-errors/src/lib.rs]: Done

File: core-backend/common/src/lib.rs [hunk 6]
```diff
@@ -70,16 +67,28 @@ pub enum TerminationReason {
     Exit(ProgramId),
     Leave,
     Success,
-    #[from]
-    Trap(TrapExplanation),
     Wait(Option<u32>, MessageWaitedType),
     GasAllowanceExceeded,
+
```
Comment (breathx) [inline core-backend/common/src/lib.rs:72]:
```suggestion
```

File: core-backend/common/src/lib.rs [hunk 7]
```diff
@@ -70,16 +67,28 @@ pub enum TerminationReason {
...
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    derive_more::Display,
+    derive_more::From,
 )]
 pub enum TrapExplanation {
+    #[from]
```
Comment (breathx) [inline core-backend/common/src/lib.rs:91]: This way we'll also allow `From<TrimmedString>`
```suggestion
```

File: core-backend/sandbox/src/funcs.rs [hunk 8]
```diff
@@ -320,9 +303,9 @@ where
             let read_inheritor_id = ctx.register_read_decoded(inheritor_id_ptr);
             let inheritor_id = ctx.read_decoded(read_inheritor_id)?;
 
-            ctx.ext.exit().map_err(ActorSyscallFuncError::Core)?;
+            ctx.ext.exit()?;
```
Comment (breathx) [inline core-backend/sandbox/src/funcs.rs:306]: move above to be able drop execution due to insufficient gas before reading data from memory

File: core-backend/sandbox/src/funcs.rs [hunk 9]
```diff
@@ -1103,7 +1000,7 @@ where
     pub fn out_of_allowance(ctx: &mut Runtime<E>, _args: &[Value]) -> SyscallOutput {
         sys_trace!(target: "syscall::gear", "out_of_allowance");
 
-        ctx.err = ActorSyscallFuncError::Core(ctx.ext.out_of_allowance()).into();
+        ctx.termination_reason = ctx.ext.out_of_allowance().into_termination_reason();
```
Comment (breathx) [inline core-backend/sandbox/src/funcs.rs:1003]: sync with @ark0f who will remove `out_of_allowance` and `out_of_gas` fns from ext as unused stuff, cause they are always free and infallible

File: core-processor/src/executor.rs [hunk 10]
```diff
@@ -351,9 +351,11 @@ where
...
     let (termination, memory, ext) = match execute() {
         Ok(report) => {
-            let (mut termination, mut memory, ext) = report
-                .into_parts()
-                .map_err(SystemExecutionError::SyscallFunc)?;
+            let BackendReport {
+                termination_reason: mut termination,
+                memory_wrap: mut memory,
+                ext,
+            } = report;
```
Comment (breathx) [inline core-processor/src/executor.rs:358]: Impl fn, use here and below
```suggestion
            (mut termination, mut memory, ext) = report.into_parts();
```

File: core/src/memory.rs [hunk 11]
```diff
@@ -660,9 +662,7 @@ impl AllocationsContext {
     ///
     /// Currently running program should own this page.
     pub fn free(&mut self, page: WasmPage) -> Result<(), Error> {
-        if page > self.max_pages {
-            Err(Error::OutOfBounds)
-        } else if page < self.static_pages || !self.allocations.remove(&page) {
+        if page < self.static_pages || page > self.max_pages || !self.allocations.remove(&page) {
```
Comment (breathx) [inline core/src/memory.rs:665]:
```suggestion
        if page < self.static_pages || page >= self.max_pages || !self.allocations.remove(&page) {
```

Comment (breathx) [reply core/src/memory.rs:665]: pages start with 0 ind, so max_pages, that means amount is already out of bounds

**Review by breathx — APPROVED** [review]
👍🏻

File: core/src/memory.rs (hunk 11 above)
Comment (grishasobol) [reply core/src/memory.rs:665]: Yes but OutOfBounds means that some body want to read or write access this pages - here we wanna just free it, so this is InvalidFree in first place

File: core-processor/src/executor.rs (hunk 10 above)
Comment (grishasobol) [reply core-processor/src/executor.rs:358]: Mmm, don't see reason to do that. I've not touched BackendReport in this PR. Also I think into_parts is less informative when read code and should be used when it's really have to

File: core-backend/common/src/lib.rs (hunk 6 above)
Comment (grishasobol) [reply core-backend/common/src/lib.rs]: Done

File: core-backend/common/src/lib.rs (hunk 7 above)
Comment (grishasobol) [reply core-backend/common/src/lib.rs]: Done

File: core-backend/sandbox/src/funcs.rs (hunk 9 above)
Comment (grishasobol) [reply core-backend/sandbox/src/funcs.rs]: Done TODO

File: core-backend/sandbox/src/funcs.rs (hunk 8 above)
Comment (grishasobol) [reply core-backend/sandbox/src/funcs.rs:303]: Done

File: core/src/memory.rs (hunk 11 above)
Comment (grishasobol) [reply core/src/memory.rs]: Done

**Pull request merged**